### ISSUE TITLE
approx_percentile_cont panics because array is not ordered

### ISSUE THREAD

**Issue by the-mousaillon** [apache/datafusion#4259]
**Describe the bug**
approx_quantile_cont panics, complaining that the input to TDigest is not ordered:
```panicked at 'unsorted input to TDigest'```

I have done some digging to understand what happens and it seams to have something to do with a corruption of indexes whithin the arrow Array.

I added a simple check to see if the array is ordered, and if not to print it in the [update_batch](https://github.com/apache/arrow-datafusion/blob/929a17525917727aa99b26667e9c3a8c55bd67c3/datafusion/physical-expr/src/aggregate/approx_percentile_cont.rs#L368-L374) function
```rust
    fn update_batch(&mut self, values: &[ArrayRef]) -> Result<()> {
        let values = &values[0];
        let v = downcast_value!(values, Int64Array);
        let sorted_values = &arrow::compute::sort(values, None)?;
        let a_down = downcast_value!(sorted_values, Int64Array);

        let mut is_sorted = true;
        for i in 0..sorted_values.len() -1 {
            if a_down.value(i) > a_down.value(i+1) {
                is_sorted = false;
                break;
            }
        }
        if !is_sorted {
           for i in 0..sorted_values.len() {
            println!("s: {}, b: {}", a_down.value(i), v.value(i));
        }
        }
        let sorted_values =
            ApproxPercentileAccumulator::convert_to_ordered_float(sorted_values)?;
        self.digest = self.digest.merge_sorted_f64(&sorted_values);
        Ok(())
    }
```

The output is surprising, we see that in seemingly every case, the first value of the sorted array *"s"* should be at the end of the array.
For intance on this array :
![image](https://user-images.githubusercontent.com/36140579/202452429-3af4bf33-88c2-4580-a50f-4ee9b26dec28.png)

The weird thing is that if I recreate the array, the sort works properly and the panic goes away.
```rust
    fn update_batch(&mut self, values: &[ArrayRef]) -> Result<()> {
        let values = &values[0];
        # If I just recreate the array before the sorting, the problem goes away
        let v = downcast_value!(values, Int64Array);
        let values: Int64Array = v.iter().map(|d| d.clone()).collect();
        let values = Arc::new(values) as ArrayRef;
        let values = &values;
        let sorted_values =
            ApproxPercentileAccumulator::convert_to_ordered_float(sorted_values)?;
        self.digest = self.digest.merge_sorted_f64(&sorted_values);
        Ok(())
}
```
This makes me think that their may be some kind of index corruption within the array buffer.

I had this bug while performing an approx_percentile_cont on a GROUP BY, whithout the GROUP BY it works fine.

**To Reproduce**
Steps to reproduce:
1. Download the parquet file: [percentile_cont_bug.zip](https://github.com/apache/arrow-datafusion/files/10032416/percentile_cont_bug.zip)

2. Execute this function

```rust
fn percentile_cont_bug() ->Result<(), ()>{
    let ctx = SessionContext::new();
    let pq_path = "...path to the provided parquet file";
    ctx.register_parquet("example", pq_path, ParquetReadOptions::default())
        .await
        .map_err(|e|())?;
    
    let q = format!("
        SELECT 
            country,
            approx_median(age) as median_age,
        FROM example
        group by 
            country
    ");
    let df = ctx.sql(&q).await
    .map_err(|e| println!("err: {:?}", e))?;
    // execute and print results
    df.show().await;
    Ok(())
}
```

**Additional context**
datafusion version: 14.0.0
arrow version: 26.0
platform: WSL (ubuntu)

**Comment by the-mousaillon** [comment]
Also, I only have this issue if I use approx_percentile_cont on an INT64 column

**Comment by HaoYang670** [comment]
Thank you @the-mousaillon for reporting the bug. We will debug it carefully. 
Also, welcome to contribute your fix to Datafusion!

**Comment by HaoYang670** [comment]
I can reproduce the bug printed here.

**Comment by HaoYang670** [comment]
cc @tustvold, could you please take a look?

**Comment by tustvold** [comment]
I can add it to my list, although I'm not familiar with the code in question. Perhaps @domodwyer or @crepererum may have some ideas

**Comment by HaoYang670** [comment]
I've tested again and @tustvold I guess you are right.

**Comment by domodwyer** [comment]
Yeah I don't remember doing anything with NULL masks when I wrote this.

**Comment by HaoYang670** [comment]
Hi @domodwyer, could you please to fix this?

**Comment by HaoYang670** [comment]
BTW, the estimate quantile algorithm doesn't follow the `paper`, any reason for this?
https://github.com/apache/arrow-datafusion/blob/df8aa7a2e2a6f54acfbfed336b84144256fb7ff8/datafusion/physical-expr/src/aggregate/tdigest.rs#L523-L524

![image](https://user-images.githubusercontent.com/59198230/203706553-91202460-aabc-4623-93ac-c16fdd7622f6.png)

**Comment by domodwyer** [comment]
Hi @HaoYang670,

I do not have time to fix this in the short term as I am working on a project. I can take a look in a few weeks - feel free to fix this yourself if you need it sooner :+1:

**Comment by Jefffrey** [comment]
> BTW, the estimate quantile algorithm doesn't follow the `paper`, any reason for this? https://github.com/apache/arrow-datafusion/blob/df8aa7a2e2a6f54acfbfed336b84144256fb7ff8/datafusion/physical-expr/src/aggregate/tdigest.rs#L523-L524
> 
> ![image](https://user-images.githubusercontent.com/59198230/203706553-91202460-aabc-4623-93ac-c16fdd7622f6.png)

I notice this wasn't ever answered: it was part of the original code which in turn was ported from another repo, see:

https://github.com/MnO2/t-digest/blob/2d27fc5ae977afd95c012b080540091349151ca3/src/lib.rs#L503

If you still want to follow up on this can create a new issue.

@HaoYang670

**Comment by HaoYang670** [comment]
@Jefffrey Thanks for catching up this old issue. Feel free to go ahead. I may not have time to follow this for now

**Comment by Jefffrey** [comment]
> [@Jefffrey](https://github.com/Jefffrey) Thanks for catching up this old issue. Feel free to go ahead. I may not have time to follow this for now

Raised #17803 to track, feel free to add some details when possible :pray:
